src/Generator/TypeGenerator/CompositeType.php

```diff
@@ -0,0 +1,157 @@
+<?php
+
+namespace Laminas\Code\Generator\TypeGenerator;
+
+use Laminas\Code\Generator\Exception\InvalidArgumentException;
+
+use function array_diff_key;
+use function array_filter;
+use function array_flip;
+use function array_map;
+use function assert;
+use function explode;
+use function implode;
+use function sprintf;
+use function str_contains;
+use function str_ends_with;
+use function str_starts_with;
+use function substr;
+use function usort;
+
+/**
+ * Represents a union/intersection type, as supported by PHP.
+ * This means that this object can be composed of {@see AtomicType} or other {@see CompositeType} objects.
+ *
+ * @internal the {@see CompositeType} is an implementation detail of the type generator,
+ *
+ * @psalm-immutable
+ */
+final class CompositeType implements TypeInterface
+{
+    public const UNION_SEPARATOR        = '|';
+    public const INTERSECTION_SEPARATOR = '&';
```

Comment:
Representing both union and intersection types as a single type is quite a mess: IMO not a solution we should roll with.